src/language/extension-types.md

```diff
@@ -0,0 +1,315 @@
+
+Extension types can't declare instance fields (unless [`external`][]),
+or [abstract members][].
+
+### Implements
+
+You can optionally use the `implements` clause to introduce
+a subtype relationship on an extension type, making *all* members
+of the representation type available to the extension type:
+
+```dart
+extension type NumberI(int i) 
+  implements int{
+  // ...
+}
+```
+
+An extension type can only implement its representation type
+or a supertype of its representation type.
+
+This is technically NOT an inheritance relationship
+like the one `implements` introduces for classes,
+but rather an [applicability][] relationship like that between an
+extension method and its `on` type.
+Members that apply to the supertype are applicable to the subtype,
+unless the supertype has a more specific definition for the member.
+
+#### `@redeclare`
+
+Replacing a method that shares a name with a member of the supertype
+is not an override relationship for extension types, but a *redeclaration*.
+Use the `@redeclare` annotation to tell the compiler you are *knowingly* choosing
+to use the same name as a supertype's member definition, not just accidentally
+using the same name:
+
+```dart
+class C {
+  void m() {}
+}
+
+extension type E(C c) implements C {
+  @redeclare
+  void m() {}
+}
+```
+
+If you don't use the annotation, the compiler will choose the [more specific][]
+definition wherever the name is invoked, which might not be the behavior you expect.
+
+## Usage
+
+To use an extension type, create an instance the same as you would with a class:
+
+```dart
+extension type NumberE(int value) {
+  NumberE operator +(NumberE other) =>
+      NumberE(value + other.value);
+
+  NumberE get i => value;
+  bool isValid() => !value.isNegative;
+}
+
+void testE() { 
+  var num = NumberE(1);
+}
+```
+
+Then you can invoke members on the object as you would with a class object.
+Any attempt to treat the extension type object as an object
+of the representation type, or vice versa, will result in compile time errors
+(unless that behavior is explicitly defined for the extension type,
+or it is [transparent](#transparency)).
+For example:
+
+```dart
+void testE() { 
+  var num = NumberE(1);
+  int num2 = NumberE(2); // Error: Can't assign 'NumberE' to 'int'.
+  
+  num.isValid(); // Ok: Extension member invocation.
+  num.isNegative(); // Error: 'NumberE' does not define 'int' member 'isNegative'.
+  
+  var sum = num + num; // Ok: 'NumberE' defines '+'.
+  var diff = num - num; // Error: 'NumberE' does not define 'int' member '-'.
+  var diff2 = num.value - 2; // Ok: Can access representation object with reference.
+  var sum2 = num + 2; // Error: Can't assign 'int' to parameter type 'NumberE'. 
+  
+  List<NumberE> numbers = [
+    NumberE(1), 
+    num.i, // Ok: 'i' getter returns type 'NumberE'.
+    1, // Error: Can't assign 'int' element to list type 'NumberE'.
+    1 as NumberE, // Ok: Dynamic type cast recognizes representation type.
```

Comment:
Not sure what "recognizes" means.

It's just OK, because it has static type `NumberE`.
It would also be (statically) OK to write `"nope" as NumberE`. That would fail at runtime, but in an example with multiple compile-time errors, runtime is a foreign country. I wouldn't expect the program to be about runtime behavior at all.

(Maybe split the example into two: One with errors, and one with correct usages. Then the reader will expect the correct usages to be *runtime* correct too.)

Reply:
Ok, it sounds like the `as` cast example just doesn't fit here so I'll remove it